pkg/program/shell/run_dir_unix.go

```diff
@@ -15,10 +15,32 @@ var (
 	errBadPermission = errors.New("bad permission")
 )
 
-// getSecureRunDir stats elvish-$uid under the default temp dir, creating it if
+// getSecureRunDir stats elvish under the XDG runtime dir, creating it if
+// it doesn't yet exist, and return the directory name.
+// If XDG runtime dir is not set it falls back to elvish-$uid under the
+// default temp dir.
+func getSecureRunDir() (string, error) {
```

Comment:
The `checkExclusiveAccess` call should also apply to $XDG_RUNTIME_DIR/elvish too.

In fact you can share the `Mkdir` call too. The only part that differs from $XDG_RUNTIME_DIR/elvish from $tmpdir/elvish is the path. There is no need to split the logic into multiple functions, either.

Reply:
I pushed a new version that's using the same function so it shares most of the code with the fallback temp directory code path.

The original version didn't apply the `checkExclusiveAccess` call because the [XDG spec](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html#variables) provides some of these guarantees:
> The directory MUST be owned by the user, and he MUST be the only one having read and write access to it. Its Unix access mode MUST be 0700.

But they're only mandated for the `XDG_RUNTIME_DIR` directory so I guess it makes sense to ensure them for `$XDG_RUNTIME_DIR/elvish` too.

Reply:
Do not trust the operating system to implement the XDG spec correctly.

In fact, do not trust any spec to be implemented correctly at all, unless you can check all the implementations.

Most of the time, Elvish only depends on functionalities of the *kernel*, so it's fairly easy to verify. But in this case, Elvish is depending on the *userland*. There are only a few major BSD implementations out there, but thousands of Linux distributions. The latter is impossible to check exhaustively.